### ISSUE TITLE
[ ERROR ] "Taking listener into use" 

### ISSUE THREAD

**Issue by palapiessa** [IronLanguages/ironpython2#374]
### Prerequisites

System specification:
•Robot Framework 3.0.3 (Python 2.7.13 on win32)
•Ironpython 2.7.8
•(RIDE 1.5.2.1) Possibly this can be reproduced without RIDE tool as well.

### Description

When I launch test from RIDE, error occurs which does not seem to block test execution.
This error did not appear in Ironpython 2.7.3

### Steps to Reproduce

1. Launch Robot Framework test script with command:
ipy -v "C:\\ProgramData\Anaconda2\Lib\site-packages\robot\run.py" %*

**Expected behavior:** Test is started without errors

**Actual behavior:** 
[ ERROR ] Taking listener 'C:\ProgramData\Anaconda2\lib\site-packages\robotide\contrib\testrunner\TestRunnerAgent.py:5601:False' into use failed: Importing listener 'C:\ProgramData\Anaconda2\lib\site-packages\robotide\contrib\testrunner\TestRunnerAgent.py' failed: Creating instance failed: TypeError: expected IBufferProtocol, got str
 Traceback (most recent call last):
   File "C:\ProgramData\Anaconda2\lib\site-packages\robotide\contrib\testrunner\TestRunnerAgent.py", line 112, in __init__
     self._send_pid()
   File "C:\ProgramData\Anaconda2\lib\site-packages\robotide\contrib\testrunner\TestRunnerAgent.py", line 128, in _send_pid
     self._send_socket("pid", os.getpid())
   File "C:\ProgramData\Anaconda2\lib\site-packages\robotide\contrib\testrunner\TestRunnerAgent.py", line 207, in _send_socket
     self.filehandler.flush()
   File "C:\Program Files\IronPython 2.7\Lib\socket.py", line 304, in flush
     view = memoryview(data)

### Versions

IronPython 2.7.8 2.7.8.0 on .NET 4.0.30319.42000

**Comment by slide** [comment]
It's odd that it didn't occur with IronPython 2.7.3, the same code is in socket.py in that release as is in 2.7.8. Thanks for the report, we'll look into it.

**Comment by slozier** [comment]
@slide Didn't we change the code to use socket.py recently? Before it was purely a C# implementation.

**Comment by slide** [comment]
@slozier Ah, you are correct. Thanks for the reminder.

**Comment by slide** [comment]
We basically need a memoryview that takes a string, the problem is that in C Python, memoryview can't be used on unicode, which is what strings in IronPython are. This requires some thought.

**Comment by slozier** [comment]
If we add an overload it should probably throw if it's not a bytes string? Another option might be modifying socket.py and adding a cast to bytes?

**Comment by slide** [comment]
We'll probably run into this in other areas, so I think adding an overload might be best

**Comment by slozier** [comment]
Probably quite inefficient, but this should work:
```C#
public MemoryView(string @object) : this(Bytes.Make(PythonOps.MakeByteArray(@object))) { }
```

**Comment by slide** [comment]
It looks like there are other methods that are not working even with this. For instance, a slice of a memoryview of an str is an str, not a memoryview. So, we may need to create a wrapper class for str that implements IBufferProtocol or something similar.

**Comment by slide** [comment]
It looks like our socket library needs to take a memoryview (or IBufferProtocol) for things in addition to str for `send` and friends.

**Comment by slozier** [comment]
Yes, to be correct it looks like we would need to accept: string, Unicode, or any buffer compatible object. I have no idea what types are covered by "buffer compatible object". At a minimum we need `MemoryView` overloads...

**Comment by simplicbe** [comment]
@palapiessa in general all releases can be found here:
https://github.com/IronLanguages/ironpython2/releases `2.7.5`: https://github.com/IronLanguages/ironpython2/releases/tag/ipy-2.7.5
As far as I know, the installer for 2.7.5 isn't available anymore.

**Comment by slozier** [comment]
@palapiessa Yes, the issue is new to 2.7.8 due to a change in socket implementation.

**Comment by slide** [comment]
https://github.com/IronLanguages/main/releases/tag/ipy-2.7.5

**Comment by slozier** [comment]
Looks like my ugly workaround made its way in https://github.com/IronLanguages/ironpython2/pull/396. We should check that it actually resolves the issue.

**Comment by tarlano** [comment]
Until 2.7.9 is released this change in socket.py also does the trick

view = str(memoryview(bytearray (data)).tolist ())

**Comment by slozier** [comment]
I think this one is fixed with https://github.com/IronLanguages/ironpython2/pull/439

**Comment by slide** [comment]
Fixed via #439
